File: symmetricTreeIterative.java
//time complexity O(n)
//space complexity O(w) where w is the max width of the queue

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if(root == null) return true;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root.left);
        q.add(root.right);
        while(!q.isEmpty()) {
            TreeNode tLeft = q.poll();
            TreeNode tRight = q.poll();
            if(tLeft == null && tRight == null) continue;
            if(tLeft != null && tRight == null) return false;
            if(tLeft == null && tRight != null) return false;
            if(tLeft.val != tRight.val) return false;
            if(tLeft.val == tRight.val) {
                q.add(tLeft.left);
                q.add(tRight.right);
                q.add(tLeft.right);
                q.add(tRight.left);
            }
        }
        return true;
    }
}

File: symmetric_tree_iterative.py
# Time complexity: O(n)
# Space complexity: O(n)
# Approach: BFS-queue
# put roots left and right in to queue
# run a loop until queue is empty
# check left and right nodes by popping them if they are same or not
# put lefts left, right's right, left's right and right's left into the queue

class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        from queue import Queue
        q = Queue()
        q.put(root.left)
        q.put(root.right)
        while not q.empty():
            left = q.get()
            right = q.get()
            if left == None and right == None:
                continue
            if left == None or right == None or left.val != right.val:
                return False
            q.put(left.left)
            q.put(right.right)
            q.put(left.right)
            q.put(right.left)
        return True